api/client/contextdialer.go

```diff
@@ -34,6 +36,38 @@ import (
 	"github.com/gravitational/teleport/api/utils/sshutils"
 )
 
+type dialConfig struct {
+	tlsConfig               *tls.Config
+	alpnConnUpgradeRequired bool
+	alpnConnUpgradeWithPing bool
```

Comment:
It is possible that `alpnConnUpgradeWithPing` param is true where `alpnConnUpgradeRequired` will be `false` ?

Reply:
yes. `alpnConnUpgradeWithPing` is independent of `alpnConnUpgradeRequired`